source/common/io/io_uring_worker_impl.cc

```diff
@@ -579,159 +545,119 @@ void IoUringServerSocket::onRead(Request* req, int32_t result, bool injected) {
-      // If there is error, let the handler to decide if the socket should be closed and stop
-      // submitting read request. And if the read error is zero, it means remote close, then
-      // needn't new request.
-      if (!read_error_.has_value() || read_error_.value() != 0) {
-        // Submit a read accept request for the next read.
-        submitReadRequest();
-      }
-    } else if (status_ == DISABLED) {
-      // Since error in a disabled socket will not be handled by the handler, stop submit read
-      // request if there is any error.
-      if (!read_error_.has_value()) {
-        // Submit a read accept request for the next read.
-        submitReadRequest();
-      }
+  } else if (status_ == Disabled) {
+    // Since error in a disabled socket will not be handled by the handler, stop submit read
+    // request if there is any error.
+    if (!read_error_.has_value()) {
+      // Submit a read accept request for the next read.
+      submitReadRequest();
     }
   }
 }
 
 void IoUringServerSocket::onWrite(Request* req, int32_t result, bool injected) {
   IoUringSocketEntry::onWrite(req, result, injected);
 
-  ENVOY_LOG(trace, "onWrite with result {}, fd = {}, injected = {}", result, fd_, injected);
-
-  // Only cleanup request for non-injected write request.
+  ENVOY_LOG(trace, "onWrite with result {}, fd = {}, injected = {}, status_ = {}", result, fd_,
+            injected, status_);
   if (!injected) {
     write_req_ = nullptr;
   }
 
-  // TODO(zhxie): even if there is pending write buffer, the socket will not handle on closing.
-  if (status_ == CLOSING) {
-    if (write_timeout_timer_) {
-      write_timeout_timer_->disableTimer();
-      write_timeout_timer_ = nullptr;
-    }
-    // Close if it is in closing status and can be closed.
-    if (read_req_ == nullptr && close_req_ == nullptr && cancel_read_req_ == nullptr &&
-        cancel_write_req_ == nullptr) {
-      if (keep_fd_open_) {
-        cleanup();
-      } else {
-        close_req_ = parent_.submitCloseRequest(*this);
-      }
-    }
-    return;
-  }
-
-  // If this is injected request, then notify the handler directly.
+  // Notify the handler directly since it is an injected request.
   if (injected) {
     ENVOY_LOG(trace,
               "there is a inject event, and same time we have regular write request, fd = {}", fd_);
-    WriteParam param{result};
-    io_uring_handler_.onWrite(param);
-    return;
-  }
-
-  // If result is EAGAIN, the request should be submitted again.
-  if (result <= 0 && result != -EAGAIN) {
-    WriteParam param{result};
-    io_uring_handler_.onWrite(param);
-    if (status_ == SHUTDOWN_WRITE || status_ == CLOSE_AFTER_SHUTDOWN_WRITE) {
-      parent_.submitShutdownRequest(*this, SHUT_WR);
+    // There is case where write injection may come after shutdown or close which should be ignored
+    // since the I/O handle or connection may be released after closing.
+    if (!shutdown_ && status_ != Closed) {
```

Comment:
This won't ignore the injection after shutdown if the shutdown finished early, the shutdown_ will become false.